data/transactions/logic/eval.go

```diff
@@ -4118,67 +4122,45 @@ func opJSONRef(cx *EvalContext) {
 
 	//parse json text
 	last = len(cx.stack) - 1
-	resp, err := parseJSON(cx.stack[last].Bytes)
+	parsed, err := parseJSON(cx.stack[last].Bytes)
 	if err != nil {
 		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
 		return
 	}
 
 	//get value from json
 	var stval stackValue
-	resp, ok := resp.(map[string]interface{})[key]
+	_, ok := parsed[key]
 	if !ok {
 		cx.err = fmt.Errorf("key %s not found in JSON text", key)
 		return
 	}
-	expectedType := jsonRefTypeNames[cx.program[cx.pc+1]]
+	expectedType := JSONRefType(cx.program[cx.pc+1])
 	switch expectedType {
-	case "JSONString":
-		val, ok := resp.(string)
-		if !ok {
-			cx.err = fmt.Errorf("got type %T but expected string", resp)
-			return
-		}
-		stval.Bytes = []byte(val)
-	case "JSONUint64":
-		val, ok := resp.(json.Number)
-		if !ok {
-			cx.err = fmt.Errorf("got type %T but expected uint64", resp)
-			return
-		}
-		// val is not a float
-		if strings.Contains(val.String(), ".") {
-			cx.err = fmt.Errorf("got type float64 but expected uint64")
-			return
-		}
-		// val is not negative
-		if val.String()[0] == '-' {
-			cx.err = fmt.Errorf("JSON value should be a uint64")
-			return
-		}
-		// val > max uint64
-		maxVal := strconv.FormatUint(math.MaxUint64, 10)
-		if val.String() > maxVal {
-			cx.err = fmt.Errorf("JSON value range within uint64 range")
+	case JSONString:
+		var value string // change to uint64 for JSONUint64
+		err := json.Unmarshal(parsed[key], &value)
+		if err != nil {
+			cx.err = fmt.Errorf("%v", err)
```

Comment:
nit: no need for `// change to uint64 for JSONUint64` comment

And I think `cx.err = err` is equivalent to `cx.err = fmt.Errorf("%v", err)`